controller/src/main/java/io/pravega/controller/store/stream/PersistentStreamBase.java

```diff
@@ -140,15 +140,23 @@ public String getScopeName() {
      * Update configuration at configurationPath.
      *
      * @param configuration new stream configuration.
-     * @return : future of boolean
+     * @return List of active segment in the stream if stream is not sealed, otherwise returns IllegalStateException.
      */
     @Override
-    public CompletableFuture<Boolean> updateConfiguration(final StreamConfiguration configuration) {
+    public CompletableFuture<List<Integer>> updateConfiguration(final StreamConfiguration configuration) {
         // replace the configurationPath with new configurationPath
-        return verifyState(() -> updateState(State.UPDATING)
-                .thenApply(x -> setConfigurationData(configuration))
-                .thenApply(x -> true),
-                Lists.newArrayList(State.ACTIVE));
+        return verifyState(() -> getActiveSegments().thenCompose(segments -> {
+            if (segments.isEmpty()) {
+                // Stream is sealed.
+                return FutureHelpers.failedFuture(
+                        StoreException.create(StoreException.Type.OPERATION_NOT_ALLOWED, "SEALED"));
+            } else {
+                // Otherwise update stream configuration.
+                return updateState(State.UPDATING).thenCompose(x -> {
+                    return setConfigurationData(configuration).thenApply(y -> segments);
+                });
+            }
+        }), Lists.newArrayList(State.ACTIVE));
```

Comment:
the code above is problematic as it has a race condition with a sealing operation:
After this verifies the state to be active, a sealed could have started.. 
And when this does getActiveSegments() it will get original segments and think that stream is not sealed.. 
Now the race here is -- by the time it goes on to do updateState(UPDATING) sealing may have finished so this is allowed and will succeed.. And post setting the state to updating it will actually update configuration on a stream that is already sealed. Which should not be allowed.

So after setting updating, it should perform the check if activeSegments are empty or not..
Also, Updating state would be misnomer as we need to fence, check validity, then change the state to updating or revert it to sealed.

Alternatively, if we retain the SEALED state, then once sealing completes it would set the state to SEALED.. hence updateState will fail as that state transition is disallowed..  



Also: another bug in the original code: change this to State.Active, State.UPDATING otherwise this update will not be idempotent.